### ISSUE TITLE
[Site Request] Youtube.com - Wait for a stream to go live before starting download

### ISSUE THREAD

**Issue by Flat** [yt-dlp/yt-dlp#878]
## Checklist

- [x] I'm reporting a site feature request
- [x] I've verified that I'm running yt-dlp version **2021.09.02**
- [x] I've searched the bugtracker for similar site feature requests including closed ones

## Description

Add functionality similar to https://github.com/hizkifw/youtube-stream-status that waits until a stream goes live before attempting to download. Currently this python script works well, however it does not work with member streams where it could most be utilized, especially for streams that will not be archived.

This is achieved by emulating how browsers check for the stream to go live via a heatbeat api that returns the next time to attempt the stream or update the time until live.

**Comment by pukkandan** [comment]
I was already thinking of implementing a more generalized implementation of this that can also be used by other extractors. What I was planning was to check the `live_status` key for `is_upcoming` and then keep re-extracting till it changes. The advantage of my idea is that this can be easily extended to other extractors; the disadvantage being that the full extraction process (downloading webpage and sometimes multiple API pages) has to happen for each "heartbeat"

**Comment by Flat** [comment]
I think that is a good idea for general use. It may not be good for YouTube members content specifically though as it may look like someone spamming the refresh or obviously using the download tool. The other downside I can think of is you may miss part of the beginning depending on the refresh rate/extraction time for the full page.

Maybe a general version with an api that allows for a more specific version for each extractor could be possible? I've tried working on a members version of the program I linked but eventually gave up as I'm missing something that identifies a user as authenticated for members content aside from authorization, cookies and a couple other headers.

**Comment by pukkandan** [comment]
> I've tried working on a members version of the program I linked but eventually gave up as I'm missing something that identifies a user as authenticated for members content aside from authorization, cookies and a couple other headers.

You can have a look at `YoutubeBaseIE._call_api` in youtube.py. The most important thing u need for authentication is the `SAPISID` header

**Comment by fuzxi** [comment]
Looks like https://github.com/yt-dlp/yt-dlp/commit/f2ebc5c7be57ef41d3fd5508e75c5ca26d36a0d1 closed this?

**Comment by pukkandan** [comment]
Right, I totally forgot this was an open request :D

**Comment by fuzxi** [comment]
Thanks for your work on this by the way, it's a major improvement to my archival setup.

**Comment by pukkandan** [comment]
Actually, reading the OP again, the request to use heartbeat API for this is not implemented. This can be reopened if you want to keep that part of the request.

**Comment by Flat** [comment]
I think it would be nice to have reopened, as the heartbeat API uses a reverse exponential back-off schema so it uses less requests the further you are from the live time, and more the closer you get so you are more likely to get the video as soon as it goes live and don't have to hammer the server to start.

**Comment by pukkandan** [comment]
@Flat have you tried the current implementation?

**Comment by Flat** [comment]
I have not as the first time I just glanced at the source, and thought it was just rand between min_wait and max_wait. After reading it it looks like it should work well. I will give it a few tries and see how it goes. Thank you @pukkandan

**Comment by Flat** [comment]
The current implementation seems to work great for my purposes, with the exception of premieres. I will try to get some log data for the next one I come across to help debug what's happening.
